Various Updates For Game Integration

zicklag opened the pull request:
This is a set of changes I needed while integrating with my game.

Each commit is self-contained and most of them fairly small.

---

This allowed me to make maybe the first useful Bevy script in a real game!

```ts
type Health = {
  0: number;
};
const Health: BevyType<Health> = { typeName: "punchy::damage::Health" };
type Stats = {
  max_health: i32,
  movement_speed: f32,
};
const Stats: BevyType<Stats> = { typeName: "punchy::fighter::Stats" };

export default {
  post_update() {
    const grabEvents = punchy.getItemGrabEvents();

    for (const event of grabEvents) {
      const fighter = event.fighter;

      const [health, stats] = world.get(fighter, Health, Stats);

      health[0] = stats.max_health;
    }
  },
};
```

That implements the logic of our health item in [Punchy](https://github.com/fishfolks/punchy).

Also I successfully implemented an event reader op, ( not included in this PR, this is in my game ) but it does require a different op for each different kind of event you want to read. I'm not sure how we might be able to improve this in the future, because of bevy's use of generics in the event resources, but at least it isn't too horrible to get working today on a per-game basis.

```rust
struct ItemGetGrabEvents;
impl JsRuntimeOp for ItemGetGrabEvents {
    fn js(&self) -> Option<&'static str> {
        Some(
            r#"
            if (!globalThis.punchy) {
                globalThis.punchy = {}
            }
            globalThis.punchy.getItemGrabEvents = () => {
                return bevyModJsScriptingOpSync('punchyGetItemGrabEvents')
                    .map(x => globalThis.wrapValueRef(x));
            }
            "#,
        )
    }

    fn run(
        &self,
        ctx: OpContext,
        world: &mut bevy::prelude::World,
        _args: bevy_mod_js_scripting::serde_json::Value,
    ) -> anyhow::Result<bevy_mod_js_scripting::serde_json::Value> {
        let event_resource = world.get_resource::<Events<ScriptItemGrabEvent>>().unwrap();

        with_state!(
            ctx.op_state,
            |event_readers: &mut HashMap<
                Handle<JsScript>,
                ManualEventReader<ScriptItemGrabEvent>,
            >,
             value_refs: &mut JsValueRefs| {
                let event_reader = event_readers
                    .entry(ctx.script_info.handle.clone_weak())
                    .or_default();

                let events = event_reader
                    .iter(event_resource)
                    .cloned()
                    .filter(|x| x.script_handle == ctx.script_info.handle)
                    .map(|x| {
                        let reflect: Box<dyn Reflect> = Box::new(x);
                        let refcell = Rc::new(RefCell::new(reflect));
                        let reflect_ref = ReflectValueRef::free(refcell);

                        JsValueRef {
                            key: value_refs.insert(reflect_ref),
                            function: None,
                        }
                    })
                    .collect::<Vec<_>>();

                Ok(serde_json::to_value(&events)?)
            }
        )
    }
}
```

zicklag committed:
Use HashSet to Store Active Scripts

This disallows inserting the same script twice.

zicklag committed:
Add Script Handle to `ScriptInfo`

zicklag committed:
Fix Clippy Lint

zicklag committed:
Make wrapValueRef Function Global

This allows it to be used in user-defined ops.

zicklag committed:
Make Important ECS Ops Types Public

zicklag committed:
Add get() Method to World

Allows you to get the components of a specific entity.

jakobhellermann commented on src/runtime/ops/ecs/query.rs:
```diff
@@ -48,3 +53,54 @@ pub fn ecs_world_query(
+    let mut query = EcsValueRefQuery::new(world, &components);
+    let results = query
+        .iter(world)
+        .filter(|x| x.entity == entity)
```
I think we should expose `EcsValueRefQuery::get` here for more efficiency, but I'll add that in `bevy_ecs_dynamic` myself

zicklag committed:
Use CamelCase for Stage Names in Scripts

This is more consistent with popular JavaScript code style.

jakobhellermann commented on src/runtime/ops/ecs/types.rs:
```diff
@@ -97,7 +97,14 @@ impl From<&ComponentInfo> for JsComponentInfo {
 }
 
-#[derive(Serialize)]
+#[derive(Deserialize, Serialize)]
+#[serde(untagged)]
+pub enum JsEntityOrValueRef {
+    JsEntity(JsEntity),
+    ValueRef(JsValueRef),
+}
```
I'm wondering if we can have a nicer abstraction around this, but I don't know what it is yet. Depends on how much of this we're gonna need of these. Not for this PR.

jakobhellermann replied:
maybe we should just return actual `Entity` valuerefs from `world.entities`

zicklag replied:
Oh, that's great idea. I'll try that.

zicklag replied:
Successfully got rid of this and just used a `JsValueRef`containing an `Entity` whenever we return entities.

jakobhellermann commented on types/lib.bevy.d.ts:
```diff
@@ -68,6 +68,7 @@ declare class World {
   resource<T>(type: BevyType<T>): T | null;
 
   query<Q extends QueryParameter[]>(...query: Q): QueryItem<Q>[];
+  get<Q extends QueryParameter[]>(entity: Entity, ...components: Q): MapQueryArgs<Q>;
```
I don't know yet if I prefer this signature over `get<C>(entity, component: BevyType<C>)` and `get_many` like you defined. It would be closer to the `fn get` in bevy.

zicklag replied:
Maybe if there is only one argument, we just return one component instead of an array of one component?

That way calling `const myComponent = world.get(entity, MyComponent)` will work like Bevy, but you can still do the multiple-get thing if you want, which feels nicely similar to the pattern you use when querying.

jakobhellermann replied:
That would be an option as well.
```ts
  get<C>(entity: Entity, component: BevyType<C>): C;
  get<Q extends QueryParameter[]>(entity: Entity, ...components: Q): MapQueryArgs<Q>;
```
and make the rust side handle both.

I think by current preference is separate `get` and `get_many` because these overloads can get a bit confusing when refactoring, e.g. changing `let [a, b] = world.get(entity, A, B);` to `let [a] = world.get(entity, A);` and having it break.

zicklag replied:
Good point, I'll separate it into `get` and `get_many`.

zicklag replied:
OK, now I'm not so sure `get_many`, because I just ran into the case where of the the many components existed on the entity, but the others didn't.

What should the behavior be in that case? Eventually in queries we'll be able to specify optional components explicitly, so maybe it's less confusing and more bevy-aligned to have a single `get` method on `World`, and then support this pattern for a multiple-get on a single entity, just like Bevy:

```ts
const query = world.query(Transform, Velocity);
const [transform, velocity] = query.get(myEntity);
```

zicklag committed:
Refactor World.get

- Make world.get return a single component
- Add QueryItems.get for conveniently getting the components related to
a specific entity in a query.
- Remove the JsEntity type in favor of returning ValueRefs to entities.

zicklag commented:
Just pushed an update that makes world.get return one component, and adds a convenience method on `QueryItems` so that you can get the components for a specific entity that matches a query.

I just have to figure out how to re-generate the Bevy types, because there's a TypeScript type error now that I removed the `Entity` type.

zicklag committed:
Update Type Definitions

zicklag commented:
There, just updated the type definitions.

zicklag committed:
Add .eq() Method for Value Refs

- Use this method to fix implementation of QueryItems.get().
- Make sure query item entities are wrapped in value refs.

zicklag commented:
Added an equality check method `.eq()` to value refs and fixed a couple bugs:

- in `world.query()` I had forgotten to wrap the entity now that it's a value ref
- in `QueryItems.get()` the equality check was invalid because it wasn't checking the equality of the underlying value ref.

zicklag commented:
Unfortunately I'm guessing there are going to be lots of instances where users try to compare two value refs with `==` or `===` and it's just not going to work right. So far I haven't seen a way to override the equality operator in JavaScript.

With the way I have it now, if you want to see if two entities are equal to each-other you have to do `entity1.eq(entity2)`. which isn't that bad, but I just know users are going to get caught forgetting or not knowing and just doing `entity1 == entity2`, which will return false unless both of those variables are actually the same variable I think.

**Edit:** I realized that JavaScript doesn't allow meaningful comparison between objects anyway, it just checks whether or not they are the same object, so maybe this isn't such a big foot gun as I thought.

jakobhellermann commented:
I added `EcsValueRefQuery::get(world, entity)` but I'm not sure how best to integrate it.

`world.query(...)` currently directly returns the array, and I wouldn't want `world.query(...).get(entity)` need to fetch the entire array.
`get` should be constant time, like in `bevy_ecs`.

We could
1. keep it as is, with `world.guery(...).get(entity)` having to iterate every match in the query (in rust or js), but that is a performance footgun
2. make `world.query` return an object with `Symbol.iterator` so that it behaves like a query, and fetches every item after the other. This seems very complicated and also inefficient because the iterator couldn't store lifetimes.
3. make `world.query` return an object with `Symbol.iterator` which fetches the array and returns that as an iterator.
```ts
query(...parameters) {
  return {
    [Symbol.iterator]: function () {
        const items = bevyModJsScriptingOpSync(
            "ecs_world_query",
            parameters,
        ).map(({ entity, components }) => ({
            entity: wrapValueRef(entity),
            components: components.map(wrapValueRef),
        }));
        return items[Symbol.iterator]();
    },
    get: function (entity) {
        const result = bevyModJsScriptingOpSync("ecs_world_get", entity, parameters);
        // return somehow
    }
  }
}
```
That way `for (const item of world.query(Transform))` works, but we lose all the array methods like `.filter` and `.all`.
4. make `world.query` return an object with methods `get(entity)` and `iter()` or `.list`
You now need to write
```ts
for (const item of world.query(Transform).iter()) {}
```
which seems unidiomatic for javascript.
5. have `world.query` return a proxy which lets `get` just get a single entity, and for every remaining property fetches the results array, caches it, and defers to it. High implementation complexity, but at least from a usage perspective we just return
```ts
interface Query extends Array<QueryItem<Q>> {
  get(Entity): QueryItem<Q>
}
```
6. keep `world.query` as is and add `world.get` and `world.get_many`. Easy to implement and understand, but doesn't match how bevy exposes it natively.

zicklag committed:
Optmize Query.get()

Avoid collecting full query results when only calling get() on a query.

zicklag commented:
I implemented option 5, because it had the best user experience and performance, and the implementation complexity didn't turn out that bad.

Pushed an update.

jakobhellermann committed:
fix some valuerefs
